### PR TITLE
Benchmark branch

### LINKED ISSUE #51: setuptools for python client
Would be nice to have `setup.py` script to be able to install the client.

### PR DESCRIPTION
**Pull request by felipecode**
Adding the benchmark branch.

Use the run_benchmark system to benchmark an agent.

Resolve #51 too.

### CODE REVIEW

**Comment by mvpcom** [comment]
Hey,  thanks for sharing this. Is this benchmark totally same as your paper? So, if I test my Agent it will show finally something similar to the paper results? All of the test including straight, one turn, navigation and navigation with dynamic obstacles?

**Comment by felipecode** [comment]
@mvpcom 
Yes ! It is totally the same. We will also provide the imitation learning code ( Tensorflow based)
on another repo in the next few days that uses this benchmark. Keep tuned.

File: PythonClient/carla/benchmarks/agent.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+
+from __future__ import print_function
+import abc
+
+from carla.planner.planner  import Planner
+
+class Agent(object):
+    def __init__(self,city_name, **kwargs):
+        __metaclass__ = abc.ABCMeta
```
Comment (nsubiron) [inline PythonClient/carla/benchmarks/agent.py]: Wouldn't be `self.__metaclass__`?

Comment (nsubiron) [reply PythonClient/carla/benchmarks/agent.py]: Same thing in Benchmark class.

File: PythonClient/carla/benchmarks/benchmark.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+                              'pos_x': -1,
+                              'pos_y': -1
+                              }
+
+        # Get the line for the experiment to be continued
+        self._line_on_file = self._continue_experiment(continue_experiment)
+
+        self._experiments = self._build_experiments()
+        # Create the log files and get the names
+        self._suffix_name, self._full_name = self._create_log_record(self._experiments)
```
Comment (nsubiron) [inline PythonClient/carla/benchmarks/benchmark.py]: This function call is missing a parameter, just passing pylint ;)

File: PythonClient/carla/benchmarks/benchmark.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+
+            with open(os.path.join(self._full_name,
+                                   'details_' + self._suffix_name), 'w') as rfd:
+
+                rw = csv.DictWriter(rfd, self._dict_rewards.keys())
+                rw.writeheader()
+            start_task = 0
+            start_pose = 0
+        else:
+            (start_task, start_pose) = self._get_pose_and_task(self._line_on_file)
```
Comment (nsubiron) [inline PythonClient/carla/benchmarks/benchmark.py]: `_get_pose_and_task` does not accept arguments.

File: PythonClient/carla/planner/city_track.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        return sorted(distance_vector)[0]
+
+
+
+    def get_distance_closest_node_route(self, pos, route):
+        distance = []
+
+        for node_iter in route:
+
+            if node_iter in self.graph.intersection_nodes():
```
Comment (nsubiron) [inline PythonClient/carla/planner/city_track.py]: Is it `self._map._graph.intersection_nodes()`?

File: PythonClient/carla/planner/graph.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        distance_dic = {}
+        for node_iter in self._nodes:
+            if node_iter != node:
+                distance_dic[sldist(node, node_iter)] = node_iter
+
+        distance_dic = collections.OrderedDict(
+            sorted(distance_dic.items()))
+
+        self._angles[node] = heading
+        for k, v in distance_dic.iteritems():
```
Comment (nsubiron) [inline PythonClient/carla/planner/graph.py]: Python 3 -> `dict.items()`

File: PythonClient/carla/planner/graph.py [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+
+            start_to_goal = np.array([node[0] - v[0], node[1] - v[1]])
+
+            print(start_to_goal)
+
+            self.angles[v] = start_to_goal / np.linalg.norm(start_to_goal)
```
Comment (nsubiron) [inline PythonClient/carla/planner/graph.py]: `self.angles` --> `self._angles`

File: PythonClient/carla/planner/graph.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        return self._nodes
+
+    def plot_ori(self, c):
+        from matplotlib import collections as mc
+
+        import matplotlib.pyplot as plt
+        line_len = 1
+
+        lines = [[(p[0], p[1]), (p[0] + line_len * self._angles[p][0],
+                                 p[1] + line_len * self._angles[p][1])] for p in self.nodes]
```
Comment (nsubiron) [inline PythonClient/carla/planner/graph.py]: `self.nodes` --> `self._nodes`

File: PythonClient/carla/planner/map.py [hunk 8]
```diff
@@ -19,54 +20,36 @@
...
+        # The built graph. This is the exact same graph that unreal builds. This
+        # is a generic structure used for many cases
+        self._graph = Graph(city_file)
 
-            self.worldrotation = np.array([
-                [math.cos(math.radians(self.angles[2])), -math.sin(math.radians(self.angles[2])), 0.0],
-                [math.sin(math.radians(self.angles[2])), math.cos(math.radians(self.angles[2])), 0.0],
-                [0.0, 0.0, 1.0]])
+        self._pixel_density = pixel_density
+        self._grid = Grid(self._graph, node_density)
```
Comment (nsubiron) [inline PythonClient/carla/planner/map.py]: Too many arguments.

**Comment by nsubiron** [comment]
https://github.com/carla-simulator/carla/blob/f7716f1ee6ed9f54ef9717e981f7085adf59be47/PythonClient/carla/planner/map.py#L84

`size` is not a variable.

File: PythonClient/carla/planner/map.py [hunk 9]
```diff
@@ -96,11 +88,12 @@ def get_map_lanes(self, height=None):
             return np.fliplr(np.asarray(img, dtype="int32"))
         return np.fliplr(self.map_image_lanes)
 
-    def get_position_on_map(self, world):
-        """Get the position on the map for a certain world position."""
+
+
+    def get_lane_orientation(self, world):
```
Comment (nsubiron) [inline PythonClient/carla/planner/map.py]: This has a bunch of member variables that seem to be missing.

File: PythonClient/carla/planner/map.py [hunk 10]
```diff
@@ -111,36 +104,64 @@ def get_position_on_map(self, world):
...
+        return self._grid._walls
+
+
+
+    def get_distance_closest_node(self, pos):
+
+        distance = []
+        for node_iter in self.graph.intersection_nodes():
+
+            distance.append(sldist(node_iter, pos))
```
Comment (nsubiron) [inline PythonClient/carla/planner/map.py]: Missing import for `sldist`.

File: PythonClient/carla/planner/map.py [hunk 11]
```diff
@@ -111,36 +104,64 @@ def get_position_on_map(self, world):
...
+    def get_walls(self):
+
+        return self._grid._walls
+
+
+
+    def get_distance_closest_node(self, pos):
+
+        distance = []
+        for node_iter in self.graph.intersection_nodes():
```
Comment (nsubiron) [inline PythonClient/carla/planner/map.py]: `self.graph` --> `self._graph`

File: PythonClient/carla/planner/planner.py [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def sldist(c1, c2): return math.sqrt((c2[0] - c1[0])**2 + (c2[1] - c1[1])**2)
+
+
+def signal(v1, v2):
+    return np.cross(v1, v2) / np.linalg.norm(v1) / np.linalg.norm(v2)
+
+
+class Planner(object):
```
Comment (nsubiron) [inline PythonClient/carla/planner/planner.py]: Not sure about these ones.
```
E:120,26: Instance of 'CarlaMap' has no 'get_pixel_density' member (no-member)
E:121,14: Instance of 'CityTrack' has no 'get_node_density' member (no-member)
E:136,15: Instance of 'Planner' has no 'is_away_from_intersection' member (no-member)
```

**Pull request merged**